### DISCUSSION TITLE
Is it possible to revert to a single revision of a single document?

### DISCUSSION THREAD

**Discussion by dtaalbers** [ravendb/ravendb#18628] [Q&A]
In my application we have research templates on which I have revisions enabled. Every time that I update a template, a new revision is created, which is great. I can also read the revisions (full and metadata) and show them in my application. Now I want to enable my users to restore the template (a single document) to a single and specific revision that my users can select in the application.

Is it possible to revert a single revision via the Client API? The only docs I could find is on how to [revert via the Studio and Point in time](https://ravendb.net/docs/article-page/6.0/csharp/document-extensions/revisions/revert-revisions). Any help would be appreciated!

**Comment by ayende** [comment]
Reverting a revision is a pretty complex story, depending on what exactly you are trying to do.
If you are trying to revert just the document's content, it is easy enough:

```
Item latest = session.Advanced.Revisions.Get<Item>(id, DateTime.UtcNow);
session.Store(latest, id);
session.SaveChanges();
```

Note that this will not deal with attachments, counters or time series. 
For those, there is a bit more complexity invovled.

**Reply by dtaalbers** [reply]
Luckily, it is just the document content for now. It might change in the future though. But for now I could use your suggestion, that looks easy enough indeed. The use case is simple for now. I just want to give my users the option to manage the revisions themselves. So I show them what revisions are available with timestamps. If possible, I want them to be able to compare revisions for changes and for them to be able to select a revision to revert back to when they made mistakes.

What if I want to revert to not the latest revision. For example, lets say 3 version back? How would I get that revision? The metadata of a revision holds an ID  right? Could I use that?

**Reply by ayende** [reply]
You fetch the revisions as usual, and you use that. Note that a revision has a unique change vector, which is basically its id

**Reply by dtaalbers** [reply]
Thanks. I don't see a way to delete specific revisions, so I guess that's not possible. I think revisions is not entirely made to fix the use-case I have. 

Basically, I have templates (just a document). I need to be able to do the following with those.

1. When a publish button is pressed in my application, the current layout should be stored as the current live version (not on every save). 
2. The old version needs to be hidden, but it should be able to be able to be re-published as the current live version.
3. I need to be able to show the difference between the published version and the previously published versions, to the user.
4. It should be possible to delete older version (because of storage reasons).
5. The ID needs of every template needs to be the same, for every version. This because the ID is referenced by other documents.

I thought the revisions would help me here, since I can store multiple versions under the same ID, but managing the revisions are not that easy. I think I can force a new revision when the publish button is pressed by using `ForceRevisionCreationFor` but I can't find how to delete a specific revision.

Is this still the way to go or are there other strategies that better suit my requirements?

**Comment by ayende** [comment]
Note that there is explicitly no way to delete a revision, yes.
Everything but (4) is easily done with RavenDB. 1 in particular can be done using:

```
session.Advanced.Revisions.ForceRevisionCreationFor(companyId);
```

If you need 4 for storage reasons, than supporting revision age or number of revision will be the way to go.

The id for 5 would be the document id, and the revisions would be the immutable id you get there.
